backend/src/stwo/circuit_builder.rs

```diff
@@ -0,0 +1,204 @@
+use stwo_prover::core::backend::Col;
+use stwo_prover::core::backend::ColumnOps;
+use stwo_prover::core::fields::m31::BaseField;
+use stwo_prover::core::fields::m31::M31;
+use stwo_prover::core::fields::{ExtensionOf, FieldOps};
+use stwo_prover::core::poly::circle::{CanonicCoset, CircleEvaluation};
+use stwo_prover::core::poly::BitReversedOrder;
+use stwo_prover::core::ColumnVec;
+
+pub type PowdrComponent<'a, F> = FrameworkComponent<PowdrEval<F>>;
+
+pub(crate) fn gen_stwo_circuit_trace<T, B, F>(
+    witness: Option<&[(String, Vec<T>)]>,
+    analyzed: Arc<Analyzed<T>>,
+) -> ColumnVec<CircleEvaluation<B, BaseField, BitReversedOrder>>
+where
+    T: FieldElement, //only Merenne31Field is supported, checked in runtime
+    B: FieldOps<M31> + ColumnOps<F>, // Ensure B implements FieldOps for M31
+    F: ExtensionOf<BaseField>,
+{
+    let element: Option<Vec<(String, Col<B, M31>)>> = Some(
+        witness
+            .as_ref()
+            .expect("Witness needs to be set")
+            .iter()
+            .map(|(name, values)| {
+                let values = values
+                    .iter()
+                    .map(|v| v.try_into_i32().unwrap().into())
+                    .collect();
+                (name.clone(), values)
+            })
+            .collect(),
+    );
+
+    let domain = CanonicCoset::new(analyzed.degree().ilog2()).circle_domain();
+    element
+        .map(|elements| {
+            elements
+                .iter()
+                .map(|(_, base_column)| CircleEvaluation::new(domain, base_column.clone()))
+                .collect()
+        })
+        .unwrap()
+}
+
+pub struct PowdrEval<T> {
+    analyzed: Arc<Analyzed<T>>,
+    witness_columns: BTreeMap<PolyID, usize>,
+}
+
+impl<T: FieldElement> PowdrEval<T> {
+    pub fn new(analyzed: Arc<Analyzed<T>>) -> Self {
+        let witness_columns: BTreeMap<PolyID, usize> = analyzed
+            .definitions_in_source_order(PolynomialType::Committed)
+            .flat_map(|(symbol, _)| symbol.array_elements())
+            .enumerate()
+            .map(|(index, (_, id))| (id, index))
+            .collect();
+
+        Self {
+            analyzed,
+            witness_columns,
+        }
+    }
+}
+
+impl<T: FieldElement> FrameworkEval for PowdrEval<T> {
+    fn log_size(&self) -> u32 {
+        self.analyzed.degree().ilog2()
+    }
+    fn max_constraint_log_degree_bound(&self) -> u32 {
+        self.analyzed.degree().ilog2() + 1
+    }
+    fn evaluate<E: EvalAtRow>(&self, mut eval: E) -> E {
+        assert!(
+            self.analyzed.constant_count() == 0 && self.analyzed.publics_count() == 0,
+            "Error: Expected no fixed columns nor public inputs, as they are not supported yet.",
+        );
+
+        let col_count = self.analyzed.commitment_count();
+        let mut witness_eval = Vec::with_capacity(col_count);
+        for _ in 0..col_count {
+            witness_eval.push(eval.next_interaction_mask(0, [0, 1]));
+        }
```

Comment:
This can be simplified by building a map of the values, instead of a vector:

```rust
        let witness_eval = self
            .witness_columns
            .iter()
            .map(|(poly_id, index)| (*poly_id, eval.next_interaction_mask(0, [0, 1])))
            .collect();
```

Then we only need to pass this `(poly_id -> value)` map to `to_stwo_expression`, instead of currently passing the `(poly_id -> index)` and `(index -> value)` maps

Reply:
Nice simplification! Done